Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/TestCertificateRepository.kt

```diff
@@ -0,0 +1,329 @@
+
+        val workedOnIds = mutableSetOf<TestCertificateIdentifier>()
+
+        internalData.updateBlocking {
+            val toRefresh = values
+                .filter { it.identifier == identifier || identifier == null }
+                .filter { it.isPending }
+
+            mutate {
+                toRefresh.forEach {
+                    workedOnIds.add(it.identifier)
+                    this[it.identifier] = when (it.type) {
+                        CoronaTest.Type.PCR -> (it as PCRCertificateContainer).copy(isUpdatingData = true)
+                        CoronaTest.Type.RAPID_ANTIGEN -> (it as RACertificateContainer).copy(isUpdatingData = true)
+                    }
+                }
+            }
+        }
+
+        internalData.updateBlocking {
+            Timber.tag(TAG).d("Checking for unregistered public keys.")
+
+            val refreshedCerts = values
+                .filter { workedOnIds.contains(it.identifier) }
+                .filter { !it.isPublicKeyRegistered }
+                .map { cert ->
+                    try {
+                        RefreshResult(registerPublicKey(cert))
+                    } catch (e: Exception) {
+                        Timber.tag(TAG).e(e, "Failed to register public key for %s", cert)
+                        RefreshResult(cert, e)
+                    }
+                }
+
+            mutate {
+                refreshedCerts
+                    .filter { it.error == null }
+                    .map { it.certificate }
+                    .forEach { this[it.identifier] = it }
+            }
+        }
+
+        internalData.updateBlocking {
+            Timber.tag(TAG).d("Checking for pending certificates.")
+
+            val refreshedCerts = values
+                .filter { workedOnIds.contains(it.identifier) }
+                .filter { it.isPublicKeyRegistered && it.isPending }
+                .map { cert ->
+                    try {
+                        RefreshResult(obtainCertificate(cert))
+                    } catch (e: Exception) {
+                        Timber.tag(TAG).e(e, "Failed to retrieve test certificate for %s", cert)
+                        RefreshResult(cert, e)
+                    }
+                }
+
+            mutate {
+                refreshedCerts
+                    .filter { it.error == null }
+                    .map { it.certificate }
+                    .forEach { this[it.identifier] = it }
+            }
+        }
+
+        internalData.updateBlocking {
+            val certs = values.filter { workedOnIds.contains(it.identifier) }
+
+            mutate {
+                certs.forEach {
+                    this[it.identifier] = when (it.type) {
+                        CoronaTest.Type.PCR -> (it as PCRCertificateContainer).copy(isUpdatingData = false)
+                        CoronaTest.Type.RAPID_ANTIGEN -> (it as RACertificateContainer).copy(isUpdatingData = false)
```

Comment:
it looks like the class hierarchy is sufficient to have a `copy` , would that work 
```kotlin                  
                   when (it) {
                        is PCRCertificateContainer -> it.copy()
                        is RACertificateContainer -> it.copy()
                    }
```